### DISCUSSION TITLE
TakeNote for serenityjs 3.0

### DISCUSSION THREAD

**Discussion by harivsudhan** [serenity-js/serenity-js#1634] [How to... ?] [answered]
@jan-molak 
Hi Jan- I need help in the following snippet, please see the screenshot as it says error.
<img width="1058" alt="image" src="https://user-images.githubusercontent.com/7137205/229534209-648b24e1-3ae8-4b15-bc3c-8441210b297c.png">

```typescript
        await apiActor.attemptsTo(
            TakeNote.of(
                List.of(LastResponse.body<UserDetails[]>())
                    .where(Property.at<UserDetails>().verificationCodeOrTable, endsWith(creditInterestTable))
                    .first()
                    .map(() => details => details.verificationId)
            ).as('ID'),
        );
        const question = q`${Note.of('ID').map(replace(/.*\//, ''))}`;
        const verificationID = await apiActor.answer(question);
        console.log(verificationID);
```

Could you please help in correcting the above based on Serenityjs 3.0

**Comment by jan-molak** [comment, marked as answer]
Hi there, please have a look at the ["What's new in Serenity/JS v3"](https://serenity-js.org/handbook/about/serenity-js-v3#taking-notes) to learn more about the changes to notes.

In general, I think a nice way to do it in Serenity/JS 3 should be something like this:

```typescript
interface MyNotes {
  ID: string;
}

await apiActor.attemptsTo(
    notes<MyNotes>().set(
        'ID', 
        List.of(LastResponse.body<UserDetails[]>())
            .where(List.item<UserDetails>().verificationCodeOrTable,  endsWith(creditInterestTable))
            .first()
            .verificationId
            .replace(/.*\//, '')
    ),
    Log.the(notes<MyNotes>().get('ID')
);
```

However, `List.item<UserDetails>().verificationCodeOrTable` doesn't exist yet - see #1420. 

For the time being, you might be able to get away using `filter`:

```typescript
await apiActor.attemptsTo(
    notes<MyNotes>().set(
        'ID', 
        LastResponse.body<UserDetails[]>()
            .filter(userDetails => userDetails.verificationCodeOrTable.endsWith(creditInterestTable))
            [0]
            .verificationId
            .replace(/.*\//, '')
    ),
    Log.the(notes<MyNotes>().get('ID')
);
```

If you think #1420 is useful, please feel free to upvote it and consider [supporting Serenity/JS](https://github.com/sponsors/serenity-js) to help me bring more features like this to the Community.
